### ISSUE TITLE
Support for shadow dom

### ISSUE THREAD

**Issue by isidrok** [lukeed/navaid#30]
Intercepting clicks made on anchors inside shadow roots looking for the anchor closest to the event target doesn't work since events are re-targeted to the shadow root host element.

The anchor can be found in the event path with the following code:

```js
function findA(e){
  var p = e.path || e.composedPath && e.composedPath(), a = p && p.shift();
  while(a && a.tagName !== 'A') a = p.shift();
  return a;
}
```

Note that this will only work for shadow roots configured with mode set to open since elements inside closed ones don't appear in the event path.

Can submit a PR if you agree with the proposed change

**Comment by lukeed** [comment]
Hmm... isn't the _point_ of Shadow DOM is to hide it away from the parent container? 
I don't use WC but iirc, you have to use `mode: 'open'` and manually proxy/bubble events _should_ you want them to be accessed. 

I'll have to think about this for now. Will open it up for feedback from others who've hopefully had more experience. Thanks for the suggestion!

**Comment by isidrok** [comment]
Yes, the point is encapsulation but that doesn't mean event delegation should not work.

It is true that you need to properly configure events your components manually dispatch if you want them to exit shadow root boundaries `composed:true` and _some_ native events won't cross shadow root boundaries but most of them do. 

You can read more about it here https://developers.google.com/web/fundamentals/web-components/shadowdom#events 

Thanks for considering.

**Comment by tomg7** [comment]
Please consider implementing this feature as navaid does not work with Shadow DOM (and libraries like [LitElement](https://lit-element.polymer-project.org/) that uses Shadow DOM by default) without it.

I added @isidrok's code to navaid's `click` function as `var x = e.target.closest('a') || findA(e)` and tested it in one of my LitElement based projects where I would like to replace Page.js with navaid. It works fine for me.

You might also want to take a look at the relevant [pull request](https://github.com/visionmedia/page.js/pull/475) from Page.js .

I think this feature would be really useful as Shadow DOM is supported in all modern browsers and its usage is rising.

Thanks

**Comment by zerodevx** [comment]
Hi @lukeed 

Please kindly add shadow dom support - the CE spec is natively supported by browsers since the last 3 major browser versions across vendors. By default, anchor link click events escape the shadow boundary - albeit re-targeted - and are natively handled by browsers just like normal anchors. It's therefore the remit of routers to handle this - just like normal anchors.

> Hmm... isn't the point of Shadow DOM is to hide it away from the parent container?
I don't use WC but iirc, you have to use mode: 'open' and manually proxy/bubble events should you want them to be accessed.

Going by this logic, then browsers should not treat anchor links in shadow dom like normal links, but that's not the case. If the component author wishes to squelch this behaviour, then he should do the whole `.attachShadow({ mode: 'closed' })`, capture events, prevent default, then dispatch a `CustomEvent` boilerplate dance.

The `event.composedPath()` method is [spec'ed](https://dom.spec.whatwg.org/) and the de-facto standard to retrieve the event path. `element.closest()` just "approximates" the event path on the basis that events bubble up to the parent node. IMO the implementation logic should be to use the latest, then fallback gracefully to support super old browsers.

So this:

https://github.com/lukeed/navaid/blob/9989c05ece026f2786f3582bb35ea4dc86afc574/src/index.js#L52-L54

Can be refactored to this:

```js
function findA(e) {
  var i = 0, p = e.path || e.composedPath && e.composedPath(), a = p && p[i];
  if (!p) return e.target.closest('a');
  while(a && a.tagName !== 'A') a = p[++i];
  return a;
}
function click(e) {
  if (e.ctrlKey || e.metaKey || e.altKey || e.shiftKey || e.button || e.defaultPrevented) return; 
  var x = findA(e), y = x && x.href; 
  ...
```

**Comment by ba55ie** [comment]
Yes please! :)

It's hard to find a simple and robust router that supports Shadow DOM (Web Components) and `navaid` seems to tick all the boxes.

By adding a few lines of code `navaid` will support a current [web standard](https://dom.spec.whatwg.org/#concept-shadow-root) and make ~a lot of~ some developers happy. ;)

**Comment by zerodevx** [comment]
FWIW, I'll maintain a [forked version](https://github.com/zerodevx/navaid) that supports shadow-dom anchor links, where the only change is from this:

https://github.com/lukeed/navaid/blob/b2b105abcf6c465621710c55e66117bf8f020b68/src/index.js#L53

to this:

```js
var x = e.composedPath, x = e.path || x && x() || [e.target], 
  x = x[0].closest('a'), y = x && x.getAttribute('href');
```

Install with: 

```
$ npm i @zerodevx/navaid
```
